src/rdf4cpp/util/CharMatcher.cpp

```diff
@@ -0,0 +1,155 @@
+#include "CharMatcher.hpp"
+
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "rdf4cpp/util/CharMatcher.cpp"
+#include <hwy/foreach_target.h>
+
+#include <cassert>
+#include <hwy/highway.h>
+#include <stdexcept>
+
+
+HWY_BEFORE_NAMESPACE();  // at file scope
+namespace rdf4cpp::util::char_matcher_detail::HWY_NAMESPACE {
+    // Calls `func(d, v)` for each input vector; out of bound lanes with index i >=
+    // `count` are instead taken from `no[i % Lanes(d)]`.
+    // if func returns false, exits early
+    // from hwy/contrib/algo/transform-inl.h
+    // TODO move to some header, probably will get used somewhere else
+    template<class D, class Func, typename T = hwy::HWY_NAMESPACE::TFromD<D>>
+    void Foreach(D d, T const *HWY_RESTRICT in, size_t const count, hwy::HWY_NAMESPACE::Vec<D> const no,
+                 Func const &func) {
+        size_t const N = Lanes(d);
+
+        size_t idx = 0;
+        if (count >= N) {
+            for (; idx <= count - N; idx += N) {
+                hwy::HWY_NAMESPACE::Vec<D> const v = LoadU(d, in + idx);
+                if (!func(d, v)) {
+                    return;
+                }
+            }
+        }
+
+        // `count` was a multiple of the vector length `N`: already done.
+        if (HWY_UNLIKELY(idx == count))
+            return;
+
+        size_t const remaining = count - idx;
+        HWY_DASSERT(0 != remaining && remaining < N);
+        hwy::HWY_NAMESPACE::Vec<D> const v = LoadNOr(no, d, in + idx, remaining);
+        func(d, v);
+    }
+
+    std::optional<bool> try_match_simd_impl(std::string_view data, std::array<CharRange, 3> const &ranges, std::string_view single) {
+        // register usage:
+        // 1 - data
+        // 6 - ranges
+        // 5 - logic
+        // ? - singles
+        // 16 overall
+        bool found_unicode = false;
+        bool r = true;
+        using D = hwy::HWY_NAMESPACE::ScalableTag<int8_t>;  //NOLINT
+        using V = hwy::HWY_NAMESPACE::VFromD<D>;             //NOLINT
```

Comment:
A very short comment what D and V are would be helpful, I know they say what they are in the documentation but having it here would be helpful